src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeHandle.cs

```diff
@@ -0,0 +1,249 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.ConstrainedExecution;
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    // This implementation does not employ critical execution regions and thus cannot
+    // reliably guarantee handle release in the face of thread aborts.
+
+    /// <summary>Represents a wrapper class for operating system handles.</summary>
+    public abstract partial class SafeHandle : CriticalFinalizerObject, IDisposable
+    {
+        // IMPORTANT:
+        // - Do not add or rearrange fields as the EE depends on this layout,
+        //   as well as on the values of the StateBits flags.
+        // - The EE may also perform the same operations using equivalent native
+        //   code, so this managed code must not assume it is the only code
+        //   manipulating _state.
+
+        /// <summary>Specifies the handle to be wrapped.</summary>
+        [SuppressMessage("Microsoft.Security", "CA2111:PointersShouldNotBeVisible")]
+        protected IntPtr handle;
+        /// <summary>Combined ref count and closed/disposed flags (so we can atomically modify them).</summary>
+        private volatile int _state = StateBits.RefCountOne; // Ref count 1 and not closed or disposed.
+        /// <summary>Whether we can release this handle.</summary>
+        private readonly bool _ownsHandle;
+        /// <summary>Whether constructor completed.</summary>
+        private volatile bool _fullyInitialized;
```

Comment:
I am curious when can this field be false or put it another way, do we need it at all?

Reply:
I think it can happen with optimizations in some corner cases. The RyuJIT optimizer can move operations that throws around raw object allocation. So if your code is `new SafeHandle(OperationThatMayThrow())`, the generated code may end up being:

```
allocate raw object
operation that may throw
call constructor
```

If the operation throws, you will end up with `_fullyInitialized == false` in the finalizer.

Reply:
Another case where you can end up with `_fullyInitialized` in the finalizer is when JITing of the SafeHandle constructor fails for some reason.

Reply:
That sounds that `private readonly bool _fullyInitialized;` should still work. Although I am surprised that finalizer runs if ctor fails to JIT

Reply:
Or if a derived SafeHandle's ctor throws prior to invoking base.ctor.

(Or if a thread abort occurred, which isn't a huge concern given that on coreclr they only exist in support of the debugger.)